File: src/main/java/br/com/blz/testjava/service/impl/ProductService.java
package br.com.blz.testjava.service.impl;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import br.com.blz.testjava.api.v1.model.ProductDTO;
import br.com.blz.testjava.converter.IProductConverter;
import br.com.blz.testjava.domain.Product;
import br.com.blz.testjava.exception.BusinessException;
import br.com.blz.testjava.exception.DuplicateItemException;
import br.com.blz.testjava.exception.NotFoundException;
import br.com.blz.testjava.repository.IProductRepository;
import br.com.blz.testjava.service.IProductService;

@Service
public class ProductService implements IProductService {

	@Autowired
	private IProductRepository productRepository;
	
	@Autowired
	private IProductConverter productConverter;
	
	@Override
	public ProductDTO findOne(final Long sku) {
		this.checkIfExists(sku);
		
		return this.productConverter
				.toDTO(this.productRepository.findOne(sku));
	}

	@Override
	public ProductDTO create(final ProductDTO dto) {
		this.checkDuplicateItem(dto.getSku());
		
		final Product product = this.productConverter.toDomain(dto);
		final Product newProduct =this.productRepository.create(product);
		return this.productConverter.toDTO(newProduct);
	}

	@Override
	public void delete(final Long sku) {
		this.checkIfExists(sku);
		
		this.productRepository.delete(sku);
	}

	@Override
	public ProductDTO update(final Long sku, final ProductDTO dto) {
		this.checkIfExists(sku);
		if(!sku.equals(dto.getSku())) 
			throw new BusinessException(String
					.format("O sku do produto (%s) é diferente do sku informado para alteração (%s)", dto.getSku(), sku));
		
		final Product productToUpdate = this.productConverter.toDomain(dto);
		return this.productConverter
				.toDTO(this.productRepository.update(sku, productToUpdate));
	}
	
	private void checkIfExists(final Long sku) {
		if(!this.productRepository.existsBySku(sku)) 
			throw new NotFoundException(String.format("Produto não encontrado com o sku: %s", sku));
	}
	
	private void checkDuplicateItem(final Long sku) {
		if(this.productRepository.existsBySku(sku)) 
			throw new DuplicateItemException(
					String.format("Erro ao adicionar o produto. Já existe um produto cadastrado com o sku: %s.", sku));
	}

	@Override
	public void deleteAll() {
		this.productRepository.deleteAll();
	}
}

File: src/main/java/br/com/blz/testjava/product/service/ProductService.kt
package br.com.blz.testjava.product.service

import br.com.blz.testjava.product.converters.toDTO
import br.com.blz.testjava.product.converters.toEntity
import br.com.blz.testjava.product.dto.ProductRequestDTO
import br.com.blz.testjava.product.dto.ProductResponseDTO
import br.com.blz.testjava.product.exceptions.ProductExistentException
import br.com.blz.testjava.product.exceptions.ProductNotFoundException
import br.com.blz.testjava.product.repository.ProductRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ProductService(@Autowired private val productRepository: ProductRepository) {

  /**
   * Cria um novo produto
   * caso já existe, lança uma exception
   */
  fun create(product: ProductRequestDTO) {
    validateExistent(product.sku)
    productRepository.save(product.toEntity())
  }

  /**
   * Procura o produto no repositório e atualiza,
   * caso não encontre lança uma exception
   */
  fun update(sku: Long, product: ProductRequestDTO) {
    val existent = productRepository.findBySku(sku)
    if (existent.isPresent)
      productRepository.save(product.toEntity())
    else
      throw productNotFound(sku)
  }

  /**
   * Retorna um produto atráves da SKU,
   * caso não encontre retorna uma exception
   */
  fun getBySku(sku: Long): ProductResponseDTO {
    val existent = productRepository.findBySku(sku)
    if (existent.isPresent)
      return existent.get().toDTO()
    throw productNotFound(sku)
  }

  /**
   * Recupera um produto por sku e exclui,
   * Caso não encontre retorna uma exception
   */
  fun deleteProduct(sku: Long) {
    val existent = productRepository.findBySku(sku)
    if (existent.isPresent) {
      productRepository.deleteBySku(sku)
    } else
      throw productNotFound(sku)
  }

  private fun productNotFound(sku: Long): ProductNotFoundException{
    return ProductNotFoundException("Produto $sku não encontrado no cadastro")
  }

  private fun validateExistent(sku: Long) {
    if (productRepository.findBySku(sku).isPresent) {
      throw ProductExistentException("Já existe um produto cadastrado com essa SKU: $sku")
    }
  }

}
